OSSM-8215 [DOC] Document using multiple control planes on a single cluster (#85272)
openshift/openshift-docs: modules/ossm-deploy-application-workloads-in-each-mesh.adoc

// Module included in the following assemblies:
// install/ossm-deploying-multiple-service-meshes-on-single-cluster.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-deploy-application-workloads-in-each-mesh_{context}"]
= Deploy application workloads in each mesh

Use this procedure to deploy application workloads in separate namespaces.

.Procedure

. Create an application namespace called `app-ns-1` by running the following command:
+
[source,terminal]
----
$ oc create namespace app-ns-1
----

. Add the `istio-discovery=mesh-1` label to ensure that the namespace is discovered by the first control plane by running the following command:
+
[source,terminal]
----
$ oc label namespace app-ns-1 istio-discovery=mesh-1
----

. Add the `istio.io/rev=mesh-1` label to the namespace to enable sidecar injection into all the pods by default while ensuring that pods in this namespace are mapped to the first control plane by running the following command:
+
[source,terminal]
----
$ oc label namespace app-ns-1 istio.io/rev=mesh-1
----

. Optional: You can verify the `mesh-1` revision name by running the following command:
[source,terminal]
----
$ oc get istiorevisions
----

> . Deploy the `sleep` and `httpbin` applications by running the following commands:

Author: I don't think it is an example app but essential to the step. I am leaving it as it is for now. :)

+
[source,terminal]
----
$ oc apply -n app-ns-1 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----

. Wait for the `httpbin` and `sleep` pods to run with sidecars injected by running the following command:
+
[source,terminal]
----
$ oc get pods -n app-ns-1
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-7f56dc944b-kpw2x   2/2     Running   0          2m26s
sleep-5577c64d7c-b5wd2     2/2     Running   0          91m
----

. Create two more namespaces called `app-ns-2` and `app-ns-3` by running the following command:
+
[source,terminal]
----
$ oc create namespace app-ns-2
----
+
[source,terminal]
----
$ oc create namespace app-ns-3

> ----

Author: I have made it as two steps.

. Add the label `istio-discovery=mesh-2` to both namespaces and the revision label `mesh-2` to match the discovery selector of the second control plane by running the following command:
+
[source,terminal]
----
$ oc label namespace app-ns-2 app-ns-3 istio-discovery=mesh-2 istio.io/rev=mesh-2
----

. Deploy the `sleep` and `httbin` applications to both namespaces by running the following commands:
+
[source,terminal]
----
$ oc apply -n app-ns-2 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[source,terminal]
----
$ oc apply -n app-ns-3 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----

. Optional: You can use the following command to wait for deployment to be ready:

[source,terminal]
----
$ oc wait deployments -n app-ns-2 --all --for condition=Available
----

.Verification

> After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane.

Reviewer: No non-step text in step blocks. Probably make this step 1 and add 1&2 as a&b.

> . Run the following procedure to verify workloads assigned to the control plane in `istio-system-1`:

Reviewer: Goal-first (this one and the following one)
```suggestion
. Verify the workloads assigned to the control plane in `istio-system-1` by running the following command:
```

----

. Run the following procedure to verify the workloads assigned to the control plane in `istio-system-2`:
+
[source,terminal]
----
$ istioctl ps -i istio-system-2
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                                  CLUSTER        CDS                LDS                EDS                RDS                ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-54gjs.app-ns-3     Kubernetes     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
httpbin-7f56dc944b-gnh72.app-ns-2     Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-k9mxz.app-ns-2       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-m9hvm.app-ns-3       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
----

> Use the following procedure to verify that application connectivity is restricted to workloads within their respective mesh:

Reviewer: Same here as the previous instance - reformat to avoid non-step text in a step block.

// Module included in the following assemblies:
// install/ossm-deploying-multiple-service-meshes-on-single-cluster.adoc

:_mod-docs-content-type: Procedure
[id="ossm-deploy-application-workloads-in-each-mesh_{context}"]
= Deploy application workloads in each mesh
:context: ossm-deploy-application-workloads-in-each-mesh

> This procedure demonstrates how to deploy application workloads in separate namespaces.

Reviewer 2: ISG:

> Avoid self-referential text whenever possible, such as “This topic is about” or “This section describes”. Instead, focus on the purpose or value of the content, and include words that users are likely to search for. For example, in a short description in an online help topic, provide users with enough information to determine whether the topic meets their needs.

.Procedure

> . Create an application namespace called `app-ns-1`:

Reviewer 2: See openshift-docs guidance for language that introduces commands: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command

+
[source,terminal]
----
$ oc create namespace app-ns-1
----

> . Add the `istio-discovery=mesh-1` label to ensure that the namespace is discovered by the first control plane.

Reviewer 2: Same as previous note, and remember to use a colon to introduce commands.

+
[source,terminal]
----
$ oc label namespace app-ns-1 istio-discovery=mesh-1
----

> . Add the `istio.io/rev=mesh-1` label to the namespace to enable sidecar injection into all the pods by default while ensuring that pods in this namespace are mapped to the first control plane.

Reviewer 2: See previous notes; apply them to command-containing steps beyond this point, too.

+
[source,terminal]
----
$ oc label namespace app-ns-1 istio.io/rev=mesh-1
----
+
[NOTE]
====
You can verify the `mesh-1` revision name by running the following command:
[source,terminal]
----
$ oc get istiorevisions
----

> ====

Reviewer 2: Could this be a verification step, or an explicit step in the process, even?

Author: Marked it as an optional step, the reason it was marked as a NOTE.

. Deploy the `sleep` and `httpbin` applications:
+
[source,terminal]
----
$ oc apply -n app-ns-1 \

>    -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml

Reviewer 3: I tried it, and it looks like the command is missing a "\" at the end, as only the first service was created. Adding this fixed it.
```suggestion
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
```

   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----

. Wait for the `httpbin` and `sleep` pods to run with sidecars injected:
+
[source,terminal]
----
$ oc get pods -n app-ns-1
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-7f56dc944b-kpw2x   2/2     Running   0          2m26s
sleep-5577c64d7c-b5wd2     2/2     Running   0          91m
----

. Create two more namespaces called `app-ns-2` and `app-ns-3`:
+
[source,terminal]
----
$ oc create namespace app-ns-2
----
+
[source,terminal]
----
$ oc create namespace app-ns-3
----

> . Add labels to both namespaces with `istio-discovery=mesh-2` and the revision label `mesh-2` to match the discovery selector of the second control plane:

Reviewer 2: For translation's sake, `istio-discovery=mesh-2` needs to have a noun.

Author: Rephrased to make it more clear.

+
[source,terminal]
----
$ oc label namespace app-ns-2 app-ns-3 istio-discovery=mesh-2 istio.io/rev=mesh-2
----

. Deploy the `sleep` and `httbin` applications to both namespaces:
+
[source,terminal]
----
$ oc apply -n app-ns-2 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[source,terminal]
----
$ oc apply -n app-ns-3 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[NOTE]
====
You can use the following command to wait for deployment to be ready:

[source,terminal]
----
$ oc wait deployments -n app-ns-2 --all --for condition=Available
----

> ====

Reviewer 2: Should I always do this? Should I do it under certain circumstances only?

Author: Again, marking it as an optional step.

.Verification

After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane.

> .. Run the following procedure to verify workloads assigned to the control plane in `istio-system-1`:

Reviewer 2: Bump out the items in this list by a level, i.e. `..` -> `.`

[source,terminal]
----
$ istioctl ps -i istio-system-1
----
You should see output similar to the following example:

.Example output
[source,terminal]
----
NAME                                  CLUSTER        CDS              LDS              EDS              RDS              ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-vwfm5.app-ns-1     Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
sleep-5577c64d7c-d675f.app-ns-1       Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
----

.. Run the following procedure to verify the workloads assigned to the control plane in `istio-system-2`:
[source,terminal]
----
$ istioctl ps -i istio-system-2
----
You should see output similar to the following example:

> .Example output

Reviewer 2: Use a plus sign to keep the example outputs from escaping the list.

httpbin-7f56dc944b-54gjs.app-ns-3     Kubernetes     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
httpbin-7f56dc944b-gnh72.app-ns-2     Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-k9mxz.app-ns-2       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-m9hvm.app-ns-3       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
----

Use the following procedure to verify that application connectivity is restricted to workloads within their respective mesh:

. Send a request from the `sleep` pod in `app-ns-1` to the `httpbin` service in `app-ns-2`. The communication should fail:
+
[source,terminal]
----
$ oc -n app-ns-1 exec deploy/sleep -c sleep -- curl -sIL http://httpbin.app-ns-2.svc.cluster.local:8000
----
+

> You should see output similar to the following example:

Reviewer 3: ```suggestion
The `PeerAuthentication` resources we created earlier enforce mutual TLS (mTLS) traffic with the `STRICT` mode in each mesh. As each mesh has its own root certificate, managed by the `istio-ca-root-cert` config map, they are not able to communicate with each other. Thus, you should see output similar to the following example indicating that communication failed. :
``` @luksa Maybe have a quick review of the above sentence to see if it is correct or could be improved. I thought it was important to provide a bit more explanation on why this example fails.

Reviewer 4: SGTM

// Module included in the following assemblies:
// install/ossm-deploying-multiple-service-meshes-on-single-cluster.adoc

:_mod-docs-content-type: Procedure
[id="ossm-deploy-application-workloads-in-each-mesh_{context}"]
= Deploy application workloads in each mesh
:context: ossm-deploy-application-workloads-in-each-mesh

This procedure demonstrates how to deploy application workloads in separate namespaces.

> Run the following command to check for existing {istio} instances.

Reviewer 3: As mentioned in the other comment, it looks like the `oc get istios` command (line 11-14) was not moved with the rest of the content under "Deploy multiple control planes", and is out of place here.  There is a sentence in that paragraph "Run the following command to check for existing Istio instances", which this command should come after.

[source,terminal]
----
$ oc get istios
----

.Procedure

. Create an application namespace called `app-ns-1`:
+
[source,terminal]
----
$ oc create namespace app-ns-1
----

. Add the `istio-discovery=mesh-1` label to ensure that the namespace is discovered by the first control plane.
+
[source,terminal]
----
$ oc label namespace app-ns-1 istio-discovery=mesh-1
----

> . Add the `istio.io/rev=mesh-1` label to the namespace to enable sidecar injection by default while ensuring that pods in this namespace are mapped to the first control plane.

Reviewer 4: Is it clear what "by default" means? Maybe clarify by saying "... to enable sidecar injection into _all pods_ by default", since that's what this label does (i.e. without this namespace label, users would have to enable injection for each pod individually).

+
[source,terminal]
----
$ oc label namespace app-ns-1 istio.io/rev=mesh-1
----
+
[NOTE]
====
You can verify the `mesh-1` revision name by running the following command:
[source,terminal]
----
$ oc get istiorevisions
----
====

. Deploy the `sleep` and `httpbin` applications:
+

> [source,terminal]
> ----
> $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-1
> ----
> +
> [source,terminal]
> ----
> $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml -n app-ns-1
> ----

Reviewer 4: This can be consolidated into a single command, as you did for the `app-ns-2` and `app-ns-3` namespaces:
```suggestion
[source,terminal]
----
$ oc apply -n app-ns-1 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
```

. Wait for the `httpbin` and `sleep` pods to run with sidecars injected:
+
[source,terminal]
----
$ oc get pods -n app-ns-1
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-7f56dc944b-kpw2x   2/2     Running   0          2m26s
sleep-5577c64d7c-b5wd2     2/2     Running   0          91m
----

. Create two more namespaces called `app-ns-2` and `app-ns-3`:
+
[source,terminal]
----
$ oc create namespace app-ns-2
----
+
[source,terminal]
----
$ oc create namespace app-ns-3
----

> . Add both namespaces to the second control plane with the discovery selector `istio-discovery=mesh-2` and the revision label with the value `mesh-2`:

Reviewer 4: This reads as though we're adding a discovery selector to the namespace, but that's not the case. We're adding a label _that will match the control plane's discovery selector_.

+

> [source,terminal]
> ----
> $ oc label namespace app-ns-2 istio-discovery=mesh-2 istio.io/rev=mesh-2
> ----
> +
> [source,terminal]
> ----
> $ oc label namespace app-ns-3 istio-discovery=mesh-2 istio.io/rev=mesh-2
> ----

Reviewer 4: Can be consolidated into a single command:
```suggestion
[source,terminal]
----
$ oc label namespace app-ns-2 app-ns-3 istio-discovery=mesh-2 istio.io/rev=mesh-2
----
```

. Deploy the `sleep` and `httbin` applications to both namespaces:
+
[source,terminal]
----
$ oc apply -n app-ns-2 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[source,terminal]
----
$ oc apply -n app-ns-3 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[NOTE]
====
You can use the following command to wait for deployment to be ready:

[source,terminal]
----

> $ oc wait deployments -n app-ns-1 --for condition=Available

Reviewer 4: ```suggestion
$ oc wait deployments -n app-ns-2 --for condition=Available
```

----
====

.Verification

After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane.

> .. Run the following procedure to verify the first control plane deployed in `istio-system-1`:

Reviewer 4: ```suggestion
.. Run the following procedure to verify workloads assigned to the control plane in `istio-system-1`:
```

[source,terminal]
----
$ istioctl ps -i istio-system-1
----
You should see output similar to the following example:

.Example output
[source,terminal]
----
NAME                                  CLUSTER        CDS              LDS              EDS              RDS              ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-vwfm5.app-ns-1     Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
sleep-5577c64d7c-d675f.app-ns-1       Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
----

> .. Run the following procedure to verify the first control plane deployed in `istio-system-2`:

Reviewer 4: ```suggestion
.. Run the following procedure to verify the workloads assigned to the control plane in `istio-system-2`:
```

[source,terminal]
----
$ istioctl ps -i istio-system-2
----
You should see output similar to the following example:

.Example output
[source,terminal]
----
NAME                                  CLUSTER        CDS                LDS                EDS                RDS                ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-54gjs.app-ns-3     Kubernetes     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
httpbin-7f56dc944b-gnh72.app-ns-2     Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-k9mxz.app-ns-2       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-m9hvm.app-ns-3       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
----

> Use the following procedure to verify that application connectivity is restricted to workloads within their respective user groups:

Reviewer 4: ```suggestion
Use the following procedure to verify that application connectivity is restricted to workloads within their respective mesh:
```

. Send a request from the `sleep` pod in `app-ns-1` to the `httpbin` service in `app-ns-2`. The communication should fail:
+
[source,terminal]
----

> $ oc -n app-ns-1 exec "$(oc -n app-ns-1 get pod -l app=sleep -o jsonpath={.items..metadata.name})" -c sleep -- curl -sIL http://httpbin.app-ns-2.svc.cluster.local:8000

Reviewer 4: This can be simplified:
```suggestion
$ oc -n app-ns-1 exec deploy/sleep -c sleep -- curl -sIL http://httpbin.app-ns-2.svc.cluster.local:8000
```

----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
HTTP/1.1 503 Service Unavailable
content-length: 95
content-type: text/plain
date: Wed, 16 Oct 2024 12:05:37 GMT
server: envoy
----

. Send a request from the `sleep` pod in `app-ns-2` managed by `mesh-2` to the `httpbin` service in `app-ns-3` also managed by `mesh-2`. The communication should work:
+
[source,terminal]
----

> oc -n app-ns-2 exec "$(oc -n app-ns-2 get pod -l app=sleep -o jsonpath={.items..metadata.name})" -c sleep -- curl -sIL http://httpbin.app-ns-3.svc.cluster.local:8000

Reviewer 4: Can be simplified to:
```suggestion
oc -n app-ns-2 exec deploy/sleep -c sleep -- curl -sIL http://httpbin.app-ns-3.svc.cluster.local:8000
```

// Module included in the following assemblies:
// install/ossm-deploying-multiple-service-meshes-on-single-cluster.adoc

:_mod-docs-content-type: Procedure
[id="ossm-deploy-application-workloads-in-each-mesh_{context}"]
= Deploy application workloads in each mesh
:context: ossm-deploy-application-workloads-in-each-mesh

> This procedure explains how to configure a cluster to host two control planes, each managing independent service mesh data planes. The configuration involves creating separate `{istio}` resources with unique names in independent {istio} system namespaces. Eah `{istio}` resource will use a unique `revision` name, which identifies the control planes, workloads, or namespaces associated with it. These `revision` names, applied using injection or `istio.io/rev` labels, enable sidecar injection while also indicating the corresponding control plane.

Reviewer 3: See other comment - this section (line 9 to 15) should be at the start of the procedure "Deploying multiple control planes", rather than here.

> This procedure explains how to configure a cluster to host two control planes, each managing independent service mesh data planes. The configuration involves creating separate `{istio}` resources with unique names in independent {istio} system namespaces. Eah `{istio}` resource will use a unique `revision` name, which identifies the control planes, workloads, or namespaces associated with it. These `revision` names, applied using injection or `istio.io/rev` labels, enable sidecar injection while also indicating the corresponding control plane.

Reviewer 3: ```suggestion
This procedure explains how to configure a cluster to host two control planes, each managing independent service mesh data planes. The configuration involves creating separate `{istio}` resources with unique names in independent {istio} system namespaces. Each `{istio}` resource will use a unique `revision` name, which identifies the control planes, workloads, or namespaces associated with it. These `revision` names, applied using injection or `istio.io/rev` labels determine which control plane will inject the sidecar proxy into the application pods.
```

> Each `{istio}` resource must also configure `discoverySelectors` to specify which namespaces the {istio} control plane observes. Only namespaces with labels that match the configured `discoverySelectors` can join the mesh.

Reviewer 3: ```suggestion
Each `{istio}` resource must also configure `discoverySelectors` to specify which namespaces the {istio} control plane observes. Only namespaces with labels that match the configured `discoverySelectors` can join the mesh. Additionally, discovery selectors determine which control plane creates the istio-ca-root-cert ConfigMap in each namespace, which is used to encrypt traffic between services with mutual TLS within each mesh.
``` 
Some more information taken from [here](https://github.com/istio-ecosystem/sail-operator/blob/main/docs/README.md#multiple-meshes-on-a-single-cluster) that seemed useful to note.

When adding an additional {istio} control plane to a cluster with an existing control plane, ensure that the existing `{istio}` instance has `discoverySelectors` configured to avoid overlapping with the new control plane. Run the following command to check for existing `{istio}` instances.

This procedure may be extended to support more than two control planes. The primary limitation on the number of service meshes in a single cluster is the availability of cluster resources.

[NOTE]
====
There is only one `IstioCNI` resource shared by all control planes in the cluster, and it must be updated independently.
====

[source,terminal]
----

> $ oc get istios

Reviewer 3: This should come after the sentence "Run the following command to check for existing `{istio}` instances." above.

----

.Procedure

. Create an application namespace called `app-ns-1`:
+
[source,terminal]
----
$ oc create namespace app-ns-1
----

> . Add the `usergroup=usergroup-1` label to ensure that the namespace is discovered by the first control plane and the `istio.io/rev=mesh-1` to enable sidecar injection in the namespace by default.

Reviewer 3: Based on feedback from Marko, we'll remove the "usergroup" naming as it is not an Istio/ossm concept and could confuse. I also think this step should be split up for clarity as it is combining two things into one step. So I suggest this step is changed to:

```suggestion
. Add the `istio-discovery=mesh-1` label to ensure that the namespace is discovered by the first control plane:
```
The command becomes:
```
oc label namespace app-ns-1 usergroup=usergroup-1 
```
With a new step...
1. Add the `istio.io/rev=mesh-1` label to the namespace to enable sidecar injection by default while ensuring that pods in this namespace are mapped to the first control plane.
```
oc label namespace app-ns-1 istio.io/rev=mesh-1
```

----
$ oc label namespace app-ns-1 usergroup=usergroup-1 istio.io/rev=mesh-1
----
+
[NOTE]
====
You can verify the `mesh-1` revision name by running the following command:
[source,terminal]
----
$ oc get istiorevisions
----
====

. Deploy the `sleep` and `httpbin` applications:
+
[source,terminal]
----
$ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-1
----
+
[source,terminal]
----
$ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml -n app-ns-1
----

. Wait for the `httpbin` and `sleep` pods to run with sidecars injected:
+
[source,terminal]
----
$ oc get pods -n app-ns-1
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-7f56dc944b-kpw2x   2/2     Running   0          2m26s
sleep-5577c64d7c-b5wd2     2/2     Running   0          91m
----

. Create two more namespaces called `app-ns-2` and `app-ns-3`:
+
[source,terminal]
----
$ oc create namespace app-ns-2
----
+
[source,terminal]
----
$ oc create namespace app-ns-3
----

> . Add both the namespaces to the second control plane with the discovery selector `usergroup=usergroup-2` and the revision label with the value `mesh-2`:

Reviewer 3: ```suggestion
. Add both namespaces to the second control plane with the discovery selector `istio-discovery=mesh-2` and the revision label with the value `mesh-2`:
```

+
[source,terminal]
----

> $ oc label namespace app-ns-2 usergroup=usergroup-2 istio.io/rev=mesh-2

Reviewer 3: ```suggestion
$ oc label namespace app-ns-2 istio-discovery=mesh-2 istio.io/rev=mesh-2
```

----
+
[source,terminal]
----

> $ oc label namespace app-ns-3 usergroup=usergroup-2 istio.io/rev=mesh-2

Reviewer 3: ```suggestion
$ oc label namespace app-ns-3 istio-discovery=mesh-2 istio.io/rev=mesh-2
```

----

. Deploy the `sleep` and `httbin` applications to both namespaces:
+
[source,terminal]
----
$ oc apply -n app-ns-2 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[source,terminal]
----
$ oc apply -n app-ns-3 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
----
+
[NOTE]
====
You can use the following command to wait for deployment to be ready:

[source,terminal]
----

> $ oc wait deployment/sleep -n app-ns-1 --for=condition=Available

Reviewer 4: This should wait for both (or simply all) deployments.
```suggestion
$ oc wait deployments -n app-ns-1 --for condition=Available
```
